egui_glium/src/painter.rs

```diff
@@ -1,192 +1,386 @@
+            Self::Gl120 => "#version 120\n",
+            Self::Gl140 => "#version 140\n",
+            Self::Es100 => "#version 100\n",
+            Self::Es300 => "#version 300 es\n",
         }
     }
+}
+
+impl Painter {
+    pub fn new(gl: &glow::Context) -> Painter {
+        let header = ShaderVersion::get(gl).version();
+        let v_src = header.to_owned() + VERT_SRC;
+        let f_src = header.to_owned() + FRAG_SRC;
+        // TODO error handling
+        unsafe {
+            let v = gl.create_shader(glow::VERTEX_SHADER).unwrap();
+            gl.shader_source(v, &v_src);
+            gl.compile_shader(v);
+            if !gl.get_shader_compile_status(v) {
+                panic!(
+                    "Failed to compile vertex shader: {}",
+                    gl.get_shader_info_log(v)
+                );
+            }
+
+            let f = gl.create_shader(glow::FRAGMENT_SHADER).unwrap();
+            gl.shader_source(f, &f_src);
+            gl.compile_shader(f);
+            if !gl.get_shader_compile_status(f) {
+                panic!(
+                    "Failed to compile fragment shader: {}",
+                    gl.get_shader_info_log(f)
+                );
+            }
+
+            let program = gl.create_program().unwrap();
+            gl.attach_shader(program, v);
+            gl.attach_shader(program, f);
+            gl.link_program(program);
+            if !gl.get_program_link_status(program) {
+                panic!("{}", gl.get_program_info_log(program));
+            }
+            gl.detach_shader(program, v);
+            gl.detach_shader(program, f);
+            gl.delete_shader(v);
+            gl.delete_shader(f);
+
+            let u_screen_size = gl.get_uniform_location(program, "u_screen_size").unwrap();
+            let u_sampler = gl.get_uniform_location(program, "u_sampler").unwrap();
+
+            let va = gl.create_vertex_array().unwrap();
+            let vb = gl.create_buffer().unwrap();
+            let eb = gl.create_buffer().unwrap();
+
+            debug_assert_eq!(std::mem::size_of::<Vertex>(), 20);
+
+            gl.bind_vertex_array(Some(va));
+            gl.bind_buffer(glow::ARRAY_BUFFER, Some(vb));
+
+            let a_pos_loc = gl.get_attrib_location(program, "a_pos").unwrap();
+            let a_tc_loc = gl.get_attrib_location(program, "a_tc").unwrap();
+            let a_srgba_loc = gl.get_attrib_location(program, "a_srgba").unwrap();
+
+            gl.vertex_attrib_pointer_f32(
+                a_pos_loc,
+                2,
+                glow::FLOAT,
+                false,
+                std::mem::size_of::<Vertex>() as i32,
+                0,
+            );
+            gl.enable_vertex_attrib_array(a_pos_loc);
+
+            gl.vertex_attrib_pointer_f32(
+                a_tc_loc,
+                2,
+                glow::FLOAT,
+                false,
+                std::mem::size_of::<Vertex>() as i32,
+                2 * std::mem::size_of::<f32>() as i32,
```

Comment:
we could potentially use https://crates.io/crates/memoffset here for clearer (and more future-proof) code

Reply:
That's a good idea, it would also remove the need for the size_of assertion